File: script-files/measures/outputsContinuous/error_function.js
jStat = require("../node_modules/jstat");

dists = {
    error_function: {
        measurements: {
            nonCentralMoments: function (k, h) {
                return undefined;
            },
            centralMoments: function (k, h) {
                return undefined;
            },
            stats: {
                mean: function (h) {
                    return 0;
                },
                variance: function (h) {
                    return 1 / (2 * h ** 2);
                },
                standardDeviation: function (h) {
                    return Math.sqrt(this.variance(h));
                },
                skewness: function (h) {
                    return 0;
                },
                kurtosis: function (h) {
                    return 3;
                },
                median: function (h) {
                    return dists.error_function.measurements.ppf(0.5, h);
                },
                mode: function (h) {
                    return 0;
                },
            },
        }
    }
}
console.log(dists.error_function.measurements.stats.mean(0.01))
console.log(dists.error_function.measurements.stats.variance(0.01))
console.log(dists.error_function.measurements.stats.standardDeviation(0.01))
console.log(dists.error_function.measurements.stats.skewness(0.01))
console.log(dists.error_function.measurements.stats.kurtosis(0.01))
// console.log(dists.error_function.measurements.stats.median(0.01))
console.log(dists.error_function.measurements.stats.mode(0.01))

// mean_value: 0
// variance_value: 5000
// standard_deviation_value: 70.71067811865476
// skewness_value: 0
// kurtosis_value: 3
// median_value: 0
// mode_value: 0

File: continuous/distributions/error_function.py
import scipy.stats
import math


class ERROR_FUNCTION:
    """
    Error Function distribution
    https://mathworld.wolfram.com/ErrorFunctionDistribution.html
    """

    def __init__(self, measurements):
        self.parameters = self.get_parameters(measurements)
        self.h = self.parameters["h"]

    def cdf(self, x: float) -> float:
        """
        Cumulative distribution function
        Calculated using the definition of the function
        Alternative: quadrature integration method
        """
        return scipy.stats.norm.cdf((2**0.5) * self.h * x)

    def pdf(self, x: float) -> float:
        """
        Probability density function
        Calculated using definition of the function in the documentation
        """
        return self.h * math.exp(-((self.h * x) ** 2)) / math.sqrt(math.pi)

    def get_num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)

    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.h > 0
        return v1

    def get_parameters(self, measurements) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample measurements.
        The parameters are calculated by formula.

        Parameters
        ==========
        measurements: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, length, num_bins, data

        Returns
        =======
        parameters : dict
            {"h": * }
        """
        h = math.sqrt(1 / (2 * measurements.variance))

        ## Results
        parameters = {"h": h}

        return parameters


if __name__ == "__main__":
    ## Import function to get measurements
    import sys

    sys.path.append("../measurements")
    from measurements_continuous import MEASUREMENTS_CONTINUOUS

    ## Import function to get measurements
    def get_data(direction):
        sample_distribution_file = open(direction, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        return data

    ## Distribution class
    path = "../data/data_error_function.txt"
    data = get_data(path)
    measurements = MEASUREMENTS_CONTINUOUS(data)
    distribution = ERROR_FUNCTION(measurements)

    print(distribution.get_parameters(measurements))
    print(distribution.cdf(measurements.mean))
    print(distribution.pdf(measurements.mean))
